File: front/src/app/features/profile/services/profile-service.ts
import {inject, Service} from '@angular/core';
import {HttpClient} from "@angular/common/http";
import {environment} from "../../../../environments/environment";
import {ProfileResponse} from "../models/profile-response";
import {Observable} from "rxjs";

@Service()
export class ProfileService {

  private readonly httpClient = inject(HttpClient);
  readonly path = `${environment.apiUrl}/profile`;

  /**
   * Met à jour le profil de l'utilisateur connecté. Pour `email`, `name` et `password`,
   * `null` signifie "ne pas modifier ce champ" (et non "le vider") : seuls les champs
   * non-`null` sont pris en compte côté backend.
   * @param email Email de l'utilisateur
   * @param name Nom de l'utilisateur
   * @param password Mot de passe à changer
   * @param currentPassword Mot de passe actuel, requis pour confirmer toute modification.
   */
  updateProfile$(email: string | null, name: string | null, password: string | null, currentPassword: string): Observable<ProfileResponse> {
    return this.httpClient.patch<ProfileResponse>(this.path, {email, name, password, currentPassword});
  }
}

File: back/src/main/java/com/openclassrooms/mddapi/service/ProfileService.java
package com.openclassrooms.mddapi.service;

import com.openclassrooms.mddapi.dto.request.UpdateProfileRequest;
import com.openclassrooms.mddapi.dto.response.ProfileResponse;

/**
 * Service de gestion du profil utilisateur.
 */
public interface ProfileService {

    /**
     * Récupère le profil de l'utilisateur connecté avec la liste des thèmes auxquels il est abonné.
     * @param userId l'id de l'utilisateur connecté.
     * @return le profil mappé pour envoyer au client.
     */
    ProfileResponse getProfile(Long userId);

    /**
     * Met à jour partiellement le nom et/ou l'email et/ou password de l'utilisateur connecté, après vérification du mot de passe actuel.
     * @param userId l'id de l'utilisateur connecté.
     * @param request les champs à mettre à jour (les champs null ne sont pas modifiés).
     * @throws com.openclassrooms.mddapi.exception.UserNotFoundException si l'utilisateur n'existe pas.
     * @throws com.openclassrooms.mddapi.exception.InvalidCurrentPasswordException si le mot de passe
     * actuel fourni ne correspond pas à celui enregistré.
     * @return ProfileResponse le profil mis à jour.
     */
    ProfileResponse updateProfile(Long userId, UpdateProfileRequest request);
}
